bftengine/src/bcstatetransfer/BCStateTran.hpp

```diff
@@ -600,17 +599,47 @@ class BCStateTran : public IStateTransfer {
 
   void onFetchingStateChange(FetchingState newFetchingState);
 
-  // When true: log historgrams, zero source flag and counter, and then unconditionally clear the iOcontexts
-  void finalizeSource(bool logSrcHistograms);
-
   // used to print periodic summary of recent checkpoints, and collected date while in state GettingMissingBlocks
   std::string logsForCollectingStatus(const uint64_t firstRequiredBlock);
   void reportCollectingStatus(const uint64_t firstRequiredBlock, const uint32_t actualBlockSize, bool toLog = false);
   void startCollectingStats();
 
-  // These 2 variables are used to snapshot source historgrams for GettingMissingBlocks state
-  bool sourceFlag_;
-  uint8_t sourceSnapshotCounter_;
+  ///////////////////////////////////////////////////////////////////////////
+  // Source session management
+  ///////////////////////////////////////////////////////////////////////////
+  bool shouldContinue(uint16_t replicaId);
+  void initializeSource(uint16_t replicaId);
+  void logSrcHistograms();
+  // clear the iOcontexts
+  void finalizeSource();
+
+  struct SourceSession {
+   public:
+    SourceSession(logging::Logger& logger) : logger_(logger) {}
+    SourceSession() = delete;
+    void clear() {
+      LOG_INFO(logger_, KVLOG(replicaId_, startTime_, getMonotonicTimeMilli(), sourceSnapshotCounter_));
+      replicaId_ = UINT16_MAX;
+      startTime_ = 0;
+      sourceSnapshotCounter_ = 0;
+    }
+
+    bool setNew(uint16_t replicaId) {
```

Comment:
Nit: Maybe the name `startNew` can be used as this function starts a new session.